rafx-base/src/atomic_once_cell.rs

```diff
@@ -0,0 +1,434 @@
+use core::ptr;
+use std::cell::UnsafeCell;
+use std::mem::MaybeUninit;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+const SET_ACQUIRE_FLAG: usize = 1 << 1;
+const SET_RELEASE_FLAG: usize = 1 << 0;
+const IS_INIT_BITMASK: usize = SET_ACQUIRE_FLAG | SET_RELEASE_FLAG;
+
+pub struct AtomicOnceCell<T> {
+    data: MaybeUninit<UnsafeCell<T>>,
+    is_initialized: AtomicUsize,
+}
+
+impl<T> AtomicOnceCell<T> {
+    pub fn new() -> Self {
+        Self {
+            data: MaybeUninit::uninit(),
+            is_initialized: AtomicUsize::new(0),
+        }
+    }
+
+    #[inline(always)]
+    fn start_set(&self) {
+        match self
+            .is_initialized
+            .fetch_update(Ordering::AcqRel, Ordering::Relaxed, |atomic_val| {
+                Some(atomic_val | SET_ACQUIRE_FLAG)
+            }) {
+            Ok(atomic_val) => {
+                if atomic_val & IS_INIT_BITMASK > 0 {
+                    // SAFETY: Panic if multiple attempts to initialize the same index occur.
+                    panic!("cannot be set more than once");
+                }
+            }
+            _ => unreachable!(),
+        };
+    }
+
+    #[inline(always)]
+    fn end_set(&self) {
+        match self
+            .is_initialized
+            .fetch_update(Ordering::Release, Ordering::Relaxed, |atomic_val| {
```

Comment:
Comments to justify the ordering chosen here would be good

Reply:
Added a comment:
>  Use `Release` to end the protected section.